test/gtest/ucs/test_datatype.cc

```diff
@@ -219,25 +220,40 @@ UCS_TEST_F(test_datatype, hlist_for_each_extract_if) {
     ASSERT_EQ(0ul, v.size());
 
     /* add 3 elements */
-    ucs_hlist_add_tail(&head, &elem2.hlist);
-    ucs_hlist_add_head(&head, &elem1.hlist);
-    ucs_hlist_add_tail(&head, &elem3.hlist);
+    ucs_hlist_add_tail(&head, &v_elems[1]->hlist);
+    ucs_hlist_add_head(&head, &v_elems[0]->hlist);
+    ucs_hlist_add_tail(&head, &v_elems[2]->hlist);
 
     /* iterate and extract 2 elements */
     v.clear();
-    ucs_hlist_for_each_extract_if(elem, &head, hlist, elem->i < 3) {
+    ucs_hlist_for_each_extract_if(elem, &head, hlist, elem->i < 2) {
         v.push_back(elem->i);
     }
     ASSERT_EQ(2ul, v.size());
-    EXPECT_EQ(1, v[0]);
-    EXPECT_EQ(2, v[1]);
+    EXPECT_EQ(0, v[0]);
+    EXPECT_EQ(1, v[1]);
     /* iterate and extract last element */
     ucs_hlist_for_each_extract_if(elem, &head, hlist, elem->i < 100) {
         v.push_back(elem->i);
     }
-    EXPECT_EQ(3, v[2]);
+    EXPECT_EQ(2, v[2]);
+
+    EXPECT_TRUE(ucs_hlist_is_empty(&head));
+
+    /* add 3 elements */
+    ucs_hlist_add_tail(&head, &v_elems[0]->hlist);
+    ucs_hlist_add_tail(&head, &v_elems[1]->hlist);
+    ucs_hlist_add_tail(&head, &v_elems[2]->hlist);
+
+    /* iterate and delete all the extracted elements */
+    ucs_hlist_for_each_extract_if(elem, &head, hlist, true) {
+        EXPECT_EQ(v_elems[0]->i, elem->i);
+        delete v_elems[0];
```

Comment:
1. shouldn't this be delete elem?
2. i guess that would fail with valgrind? maybe also memset(v_elems[0], 0xFF, sizeof elem_t)

Reply:
> 1. shouldn't this be delete elem?
> 2. i guess that would fail with valgrind? maybe also memset(v_elems[0], 0xFF, sizeof elem_t)

1. is actually the same, i will fix to make test case clearer
2. to get segfault in case of failure without valgrind?

Reply:
2. yes